### PR TITLE
Disallow external fields without _dev/build/build.yml file

### LINKED ISSUE #510: [Change Proposal] Disallow to use external fields if there is no build.yml
Proposed in https://github.com/elastic/kibana/issues/155057#issuecomment-1516029993, related to https://github.com/elastic/elastic-package/issues/1224.

External fields are allowed even if there is no build.yml. Fail validation if external is used in a package without build.yml file.

This will also help avoiding built packages to have unresolved external fields, because packages built with elastic-package don't contain development and build files.

### PR DESCRIPTION
**Pull request by mrodm**
## What does this PR do?

Add a new semantic validation to avoid fields with the external key defined if there is no `_dev/build/build.yml` file (where the ECS reference is defined).

If any field has defined the `external` key, this new validation rule:
- checks that `_dev/build/build.yml` exists
- checks that `_dev/build/build.yml` contains a definition for the external value set in the definition:
    - currently, the only supported value is `ecs`

Example:
1. Go to some package that contains external fields (e.g. elastic_package_registry)
2. Remove `_dev/build/build.yml` file
3. Run `elastic-package build -v`
4. First, it fails becuase it cannot render the README file due to the `{{ fields metrics }}`
5. Remove that line from `_dev/build/docs/README.md`
   ```diff
   diff --git packages/elastic_package_registry/_dev/build/docs/README.md packages/elastic_package_registry/_dev/build/docs/README.md
   index 84a915bd8..c7190745f 100644
   --- packages/elastic_package_registry/_dev/build/docs/README.md
   +++ packages/elastic_package_registry/_dev/build/docs/README.md
   @@ -45,7 +45,6 @@ Elastic Package Registry can provide Prometheus metrics in the `/metrics` endpoi
    You can verify that metrics endpoint is enabled by making an HTTP request to
    `http://localhost:9000/metrics` on your package registry instance.
    
   -{{ fields "metrics" }}
    
    #### Example
    
   ```
7. Run `elastic-package build -v`
```shell
 $ elastic-package build -v
2023/04/26 16:17:00 DEBUG Enable verbose logging
...
Error: building package failed: invalid content found in built zip package: found 8 validation errors:
   1. file "/home/mariorodriguez/Coding/work/integrations/build/packages/elastic_package_registry-0.0.6.zip/data_stream/metrics/fields/ecs.yml" is invalid: field ecs.version with external key defined ("ecs") but no _dev/build/build.yml found
   2. file "/home/mariorodriguez/Coding/work/integrations/build/packages/elastic_package_registry-0.0.6.zip/data_stream/metrics/fields/ecs.yml" is invalid: field error.message with external key defined ("ecs") but no _dev/build/build.yml found
   3. file "/home/mariorodriguez/Coding/work/integrations/build/packages/elastic_package_registry-0.0.6.zip/data_stream/metrics/fields/ecs.yml" is invalid: field event.duration with external key defined ("ecs") but no _dev/build/build.yml found
   4. file "/home/mariorodriguez/Coding/work/integrations/build/packages/elastic_package_registry-0.0.6.zip/data_stream/metrics/fields/ecs.yml" is invalid: field event.ingested with external key defined ("ecs") but no _dev/build/build.yml found
   5. file "/home/mariorodriguez/Coding/work/integrations/build/packages/elastic_package_registry-0.0.6.zip/data_stream/metrics/fields/ecs.yml" is invalid: field event.kind with external key defined ("ecs") but no _dev/build/build.yml found
   6. file "/home/mariorodriguez/Coding/work/integrations/build/packages/elastic_package_registry-0.0.6.zip/data_stream/metrics/fields/ecs.yml" is invalid: field service.address with external key defined ("ecs") but no _dev/build/build.yml found
   7. file "/home/mariorodriguez/Coding/work/integrations/build/packages/elastic_package_registry-0.0.6.zip/data_stream/metrics/fields/ecs.yml" is invalid: field service.type with external key defined ("ecs") but no _dev/build/build.yml found
   8. file "/home/mariorodriguez/Coding/work/integrations/build/packages/elastic_package_registry-0.0.6.zip/data_stream/metrics/fields/ecs.yml" is invalid: field tags with external key defined ("ecs") but no _dev/build/build.yml found

```

**NOTE** This fails if it is tested in input packages. There is already an issue for that https://github.com/elastic/package-spec/issues/511

## Why is it important?

The aim of this validation is:
- Ensure that all fields are resolved in the built packages
- If there is any field not resolved (e.g. they include `external: ecs`), those packages should have the development folder. They are not built packages.

## Checklist

- [x] I have added test packages to [`test/packages`](https://github.com/elastic/package-spec/tree/main/test/packages) that prove my change is effective.
- [x] I have added an entry in [`spec/changelog.yml`](https://github.com/elastic/package-spec/blob/main/spec/changelog.yml).
- [x] Updated current packages to contain a valid `_dev/build/build.yml` file.

## Related issues

- Closes #510

### CODE REVIEW

File: code/go/internal/validator/semantic/validate_external_fields_with_dev_folder.go [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+	}
+	return validateFields(fsys, validateFunc)
+}
+
+func readDevBuildDependencies(f pkgpath.File) ([]string, error) {
+	vals, err := f.Values("$.dependencies")
+	if err != nil {
+		return []string{}, fmt.Errorf("can't read dependencies: %w", err)
+	}
+
```
Comment (jsoriano) [inline code/go/internal/validator/semantic/validate_external_fields_with_dev_folder.go:62]: Nit. I think it'd be better to check here if the value returned is what we expect.
```suggestion
	dependencies, ok := vals.(map[string]interface{})
	if !ok {
		return nil, fmt.Errorf("dependencies expected to be a map, found %T: %s", vals, vals)
	}
```

Then below we can continue working with the map already converted.

File: code/go/pkg/validator/validator_test.go [hunk 2]
```diff
@@ -432,6 +433,88 @@ func TestValidateWarnings(t *testing.T) {
...
+	}
+
+	pkgRootPath := filepath.Join("..", "..", "..", "..", "test", "packages", pkgName)
+	devFolderPath := filepath.Join(pkgRootPath, "_dev")
+	buildFolderPath := filepath.Join(devFolderPath, "build")
+	buildFilePath := filepath.Join(buildFolderPath, "build.yml")
+
+	defer func() {
+		os.RemoveAll(devFolderPath)
+	}()
```
Comment (jsoriano) [inline code/go/pkg/validator/validator_test.go:475]: Nit. It is a bit of an anti-pattern to write and remove files from the working copy. Could we do this in a temporary directory? Even if we also have to copy the rest of the package.

**Review by jsoriano — DISMISSED** [review]
Looks good, added a couple of nitpicks.

It'd be also good to test this branch with integrations to early detect if there is some package that would fail with these changes.

File: code/go/pkg/validator/validator_test.go (hunk 2 above)
Comment (mrodm) [reply code/go/pkg/validator/validator_test.go]: Sure, changed. Copied to a temporal directory and from there the tests are performed.

**Comment by mrodm** [comment]
Triggered two builds to check if there are currently uses:
- in elastic-package https://github.com/elastic/elastic-package/pull/1244
- in integrations https://github.com/elastic/integrations/pull/6013

Fields in input packages will not validated due to https://github.com/elastic/package-spec/issues/511

File: code/go/internal/validator/semantic/validate_external_fields_with_dev_folder.go (hunk 1 above)
Comment (mrodm) [reply code/go/internal/validator/semantic/validate_external_fields_with_dev_folder.go:62]: Changed.

File: code/go/internal/validator/semantic/validate_external_fields_with_dev_folder_test.go [hunk 3]
```diff
@@ -5,39 +5,51 @@
...
+			fsys := fspath.DirFS(tempDir)
+			buildFilePath := filepath.Join(tempDir, "build.yml")
+			err := os.WriteFile(buildFilePath, []byte(test.contents), 0644)
+			require.NoError(t, err)
+
+			f, err := pkgpath.Files(fsys, "build.yml")
+			require.Len(t, f, 1)
 			require.NoError(t, err)
 
+			list, err := readDevBuildDependenciesKeys(f[0])
```
Comment (jsoriano) [inline code/go/internal/validator/semantic/validate_external_fields_with_dev_folder_test.go]: Umm, as it is now, couldn't it read directly the file?
```suggestion
			buildFilePath := filepath.Join(t.TempDir(), "build.yml")
			err := os.WriteFile(buildFilePath, []byte(test.contents), 0644)
			require.NoError(t, err)

			list, err := readDevBuildDependenciesKeys(buildFilePath)
```

**Review by jsoriano — APPROVED** [review]

File: code/go/internal/validator/semantic/validate_external_fields_with_dev_folder_test.go (hunk 3 above)
Comment (mrodm) [reply code/go/internal/validator/semantic/validate_external_fields_with_dev_folder_test.go]: `readDevBuildDependenciesKey` requires as a parameter a `pkgpath.File`, I think it should be kept with this setup logic.

**Review by jsoriano — APPROVED** [review]

**Review by jsoriano — DISMISSED** [review]

**Review by jsoriano — APPROVED** [review]

**Pull request merged**